tests/unit-test.el

```diff
@@ -141,31 +141,132 @@
-            :to-be nil))
+    (cljr--with-clojure-temp-file "foo.cljs"
+      (insert "(ns foo)")
+      (expect (cljr--unresolved-alias-ref "js")
+              :to-be nil)))
 
   (it "returns js for js alias in clj file."
-    (expect (cljr--with-clojure-temp-file "foo.clj"
-              (insert "(ns foo)")
-              (cljr--unresolved-alias-ref "js"))
-            :to-equal "js"))
+    (cljr--with-clojure-temp-file "foo.clj"
+      (insert "(ns foo)")
+      (expect (cljr--unresolved-alias-ref "js")
+              :to-equal "js")))
 
   (it "returns js for js alias in cljc file."
-    (expect (cljr--with-clojure-temp-file "foo.cljc"
-              (insert "(ns foo)")
-              (cljr--unresolved-alias-ref "js"))
-            :to-equal "js")))
+    (cljr--with-clojure-temp-file "foo.cljc"
+      (insert "(ns foo)")
+      (expect (cljr--unresolved-alias-ref "js")
+              :to-equal "js"))))
+
+(defmacro with-point-at (text &rest body)
+  (declare (indent 1))
+  `(progn (insert ,text)
+          (goto-char (point-min))
+          (re-search-forward "|")
+          (delete-char -1)
+          ,@body))
+
+(describe "cljr--language-context-at-point"
+  (it "identifies a clj file"
+    (cljr--with-clojure-temp-file "foo.clj"
+      (insert "(ns foo)")
+      (expect (cljr--language-context-at-point)
+              :to-equal '("clj" nil))))
+
+  (it "identifies a cljs file"
+    (cljr--with-clojure-temp-file "foo.cljs"
+      (insert "(ns foo)")
+      (expect (cljr--language-context-at-point)
+              :to-equal '("cljs" nil))))
+
+  (it "identifies a cljc file"
+    (cljr--with-clojure-temp-file "foo.cljc"
+      (insert "(ns foo)")
+      (expect (cljr--language-context-at-point)
+              :to-equal '("cljc" nil))))
+
+  (it "identifies a cljc file with a cljs context"
+    (cljr--with-clojure-temp-file "foo.cljc"
+      (with-point-at "(ns foo) #?(:cljs (Math/log|))"
+        (expect (cljr--language-context-at-point)
+                :to-equal '("cljc" "cljs")))))
+
+  (it "identifies a cljc file with a clj context"
+    (cljr--with-clojure-temp-file "foo.cljc"
+      (with-point-at "(ns foo) #?(:clj (Math/log|))"
+        (expect (cljr--language-context-at-point)
+                :to-equal '("cljc" "clj"))))))
+
+(describe "cljr--prompt-or-select-libspec"
+  (it "prompts user for namespace selection"
+    (spy-on 'completing-read :and-return-value "[a.a :as a]")
+    (expect (cljr--prompt-or-select-libspec '("[a.a :as a]"
+                                              "[a.b :as b]"))
+            :to-equal "[a.a :as a]"))
+
+  (it "returns user content from prompt regardless of candidates"
+    (spy-on 'completing-read :and-return-value "[alpha :as a]")
+    (expect (cljr--prompt-or-select-libspec
+             '("[a.a :as a]"
+               "[a.b :as b]"))
+            :to-equal "[alpha :as a]"))
+
+  (it "short-circuits if only one candidate matches"
+    (expect 'completing-read :to-have-been-called-times 0)
```

Comment:
Is it possible to add a counterexample where completing-read is called `> 0` times?

That would make this test more logically robust.

Reply:
That's what the other three tests are doing that are replacing the return value for 'completing-read with a `spy-on` with a `:and-return-value` mock return value. The reason this particular example is asserting it's not called is because it's one of the few cases it's not in the call graph. There may be a counter-example I missed though, so I would be happy to add one if I misunderstood, but I need more explanation.

Reply:
If, as a maintainer, I read `(expect 'completing-read :to-have-been-called-times 0)`, I will expect a counterexample which says `(expect 'completing-read :to-have-been-called-times <a positive number>)`, using the same construct, in a nearby test case.

Otherwise it is overassuming to expect people to detailedly understand all tests cases, and how test cases that use different constructs relate to each other.

Basically I'm suggesting to turn an implicit similarity into an explicit one.

Reply:
Turns out I misread how the count expectations work, it has to be the at the end of the test. I thought it worked like rspec where it asserts at the beginning and then checks all assertions at the end. I added call count assertions at the end of all 4 of those test cases.